grype/db/internal/gormadapter/open.go

```diff
@@ -4,37 +4,81 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"strings"
 
 	"github.com/glebarez/sqlite"
 	"gorm.io/gorm"
+
+	"github.com/anchore/grype/internal/log"
 )
 
+var commonStatements = []string{
+	`PRAGMA foreign_keys = ON`, // needed for v6+
+}
+
 var writerStatements = []string{
-	// performance improvements (note: will result in lost data on write interruptions).
-	// on my box it reduces the time to write from 10 minutes to 10 seconds (with ~1GB memory utilization spikes)
-	`PRAGMA synchronous = OFF`,
-	`PRAGMA journal_mode = MEMORY`,
-	`PRAGMA cache_size = 100000`,
-	`PRAGMA mmap_size = 268435456`, // 256 MB
+	// performance improvements (note: will result in lost data on write interruptions)
+	`PRAGMA synchronous = OFF`,     // minimize the amount of syncing to disk, prioritizing write performance over durability
+	`PRAGMA journal_mode = MEMORY`, // do not write the journal to disk (maximizing write performance); OFF is faster but less safe in terms of DB consistency
 }
 
-var readOptions = []string{
-	"immutable=1",
-	"cache=shared",
-	"mode=ro",
+var heavyWriteStatements = []string{
+	`PRAGMA cache_size = -1073741824`, // ~1 GB (negative means treat as bytes not page count); one caveat is to not pick a value that risks swapping behavior, negating performance gains
+	`PRAGMA mmap_size = 1073741824`,   // ~1 GB; the maximum size of the memory-mapped I/O buffer (to access the database file as if it were a part of the process’s virtual memory)
+}
+
+var readConnectionOptions = []string{
+	"immutable=1",  // indicates that the database file is guaranteed not to change during the connection’s lifetime (slight performance benefit for read-only cases)
+	"mode=ro",      // opens the database in as read-only (an enforcement mechanism to allow immutable=1 to be effective)
+	"cache=shared", // multiple database connections within the same process share a single page cache
 }
 
 type config struct {
-	path   string
-	write  bool
-	memory bool
+	path                      string
+	writable                  bool
+	truncate                  bool
+	allowLargeMemoryFootprint bool
+	models                    []any
+	initialData               []any
+	memory                    bool
+	statements                []string
 }
 
 type Option func(*config)
 
-func WithTruncate(truncate bool) Option {
+func WithTruncate(truncate bool, models []any, initialData []any) Option {
+	return func(c *config) {
+		c.truncate = truncate
+		if truncate {
+			c.writable = true
+			c.models = models
+			c.initialData = initialData
+			c.allowLargeMemoryFootprint = true
+		}
+	}
+}
+
+func WithStatements(statements ...string) Option {
+	return func(c *config) {
+		c.statements = append(c.statements, statements...)
+	}
+}
+
+func WithMigrate(models []any) Option {
```

Comment:
nit: I think "migrate" can be misconstrued that these models contain migration functionality, maybe this could be named `WithModels`?